Predeployed accounts rpc call

Juul-Mc-Goa opened the pull request:
# Pull Request type

- Feature
- Testing

## What is the current behavior?

The predeployed accounts addresses defined in the genesis state are printed when running  `cargo run -- --dev`  but there are no rpc methods to obtain all the predeployed accounts information.

Resolves: #1230

## What is the new behavior?

A new method `predeployedAccounts` is created in `crates/client/rpc-core/src/lib.rs` and `crates/client/rpc/src/lib.rs`. This rpc method returns a vec of `PredeployedAccount`s, a struct defined in `rpc-core/src/lib.rs` which has the following attributes:
-  `contract_address: FieldElement`,
- `contract_class: ContractClass`,
- `balance: FieldElement`,
- `private_key: Option<FieldElement>`.

A hurl test file is created at `examples/rpc/starknet/starknet_predeployed_accounts.hurl`.

## Does this introduce a breaking change?

No

## Other information

Please note:
This rpc method was made following the instructions at `docs/rpc-contributions.md`. However, I was unable to change `starknet-rpc-test`, as the value returned by `madara.get_starknet_client()` does not implement the `StarknetRpcApi` trait.

Juul-Mc-Goa commented:
does this PR solves your issue @apoorvsadana ?

apoorvsadana commented:
Thanks for the PR @Juul-Mc-Goa! Sorry for the late review, I just got back from Devconnect yesterday. I have added some comments.

Juul-Mc-Goa commented on flake.nix:
```diff
@@ -1,39 +1,30 @@
 {
   inputs = {
-    nixpkgs.url      = "github:NixOS/nixpkgs/nixos-unstable";
```
yeah no pb wih that, I guess it's just my auto-formatter that did it.

Juul-Mc-Goa commented on crates/client/rpc/src/constants.rs:
```diff
@@ -2,3 +2,16 @@
 pub const MAX_EVENTS_CHUNK_SIZE: usize = 1000;
+
+/// Class hashes whose corresponding contracts classes are "NoValidateAccount".
+pub const NO_VALIDATE_ACCOUNT_CLASS_HASHES: [&str; 2] = [
+    "0x0279d77db761fba82e0054125a6fdb5f6baa6286fa3fb73450cc44d193c2d37f",
+    "0x35ccefcf9d5656da623468e27e682271cd327af196785df99e7fee1436b6276",
+];
+/// Predeployed accounts adresses defined in the genesis state.
+pub const PREDEPLOYED_ACCOUNTS_ADDRESSES: [&str; 4] = ["0x1", "0x2", "0x3", "0x4"];
+
+/// Pk for the relevant predeployed accounts defined in the genesis state.
+pub const ARGENT_PK: &str = "0x00c1cf1490de1352865301bb8705143f3ef938f97fdf892f1090dcb5ac7bcd1d";
+/// Fee token contract address defined in the genesis state.
+pub const FEE_TOKEN_ADDRESS: &str = "0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7";
```
Alright, but idk where such a function should be implemented.
`print_development_accounts` only has access to a GenesisLoader (which itself needs a base path to find the genesis assets), while `Starknet<...>` in the file `client/rpc/src/lib` doesn't have access to a path towards `genesis.json`, but can use  RPCs to query the state.

apoorvsadana replied:
Hmm, you're right. Actually, the idea of reading the pre funded accounts as specific indexes in the genesis state doesn't make sense. If someone forks Madara and changes the genesis state, this would break. Let's move these variables to constants and add a utils function to load them. We can add the utils function to the client itself or create a utils directory on the root level. I think the latter existed before and was removed for some reason. @tdelabro what do you think?

tdelabro replied:
This whole thing is super hacky anyway.
I think the easiest thing we can do is to return the GenesisConfig struct, without the whole nice printing around it

tdelabro replied:
Also, is this thing required? 
The printing of the deployed account when running in dev mode is a cool nice to have for developers trying to add TX to the network. But what is the usecase of having an RPC to do so? Who asked for it?

tdelabro replied:
My point is that if nobody needs it and the implementation has to be super hacky, don't do it.

apoorvsadana replied:
Braavos needs it. They call this endpoint to get the available accounts and accordingly use it for faucets. Katana and devnet already support this #1230

apoorvsadana replied:
> I think the easiest thing we can do is to return the GenesisConfig struct, without the whole nice printing around it

You mean return the GenesisConfig from the pallet to the client when an RPC call is made and remove the printing on start?

tdelabro replied:
We can easily have the list of deployed accounts by iterating over the storage.
If we do this at block 0 we have the predeployed accounts.
But we won't get information about "this one is argent, this one is Cairo0NoValidate". Is this fine?

tdelabro replied:
The whole point of the genesisloader was to not have this in a hardcoded way.

tdelabro replied:
If we embrace the genesis loader, we should do it entirely.
We can complexity the struct by adding for which contract it contains a `name` and optional `private_key` field.
This way we can read it from disk, return it and anybody can display or use it the way he want

apoorvsadana replied:
Fair, but I don't think we should add extra fields to the contracts section. Can we add a new field to `GenesisData` called `prefunded_accounts` where the user can explicitly state the `address` and the `pk`.

tdelabro replied:
```
pub struct GenesisData {
    pub contract_classes: Vec<(ClassHash, ContractClass)>,
    pub contracts: Vec<(ContractAddress, ClassHash)>,
    pub storage: Vec<(ContractStorageKey, StorageValue)>,
    pub fee_token_address: ContractAddress,
    pub seq_addr_updated: bool,
}
```

Why couldn't we add them to the `contracts` tuple?

tdelabro replied:
`    pub contracts: Vec<(ContractAddress, ClassHash, Option<String>, Option<Vec<u8>>)>,`

apoorvsadana replied:
I think it tightly couples the `contracts` information with prefunded accounts. How do I add a pre funded account which I don't want to be displayed to the user? Also, if I have very little prefunded accounts (as many chains will have in production), it doesn't make sense for me to repeatedly write `false` or `vec![]`

apoorvsadana replied:
@Juul-Mc-Goa let's add `prefunded_accounts` to the `GenesisData` in that case. We can return the genesis information from the pallet for the client to use. Will you be able to pick this?

tdelabro replied:
> We can return the genesis information from the pallet for the client to use.

I'm not sure about this. I think the file has to be read again.

Juul-Mc-Goa replied:
> @Juul-Mc-Goa let's add `prefunded_accounts` to the `GenesisData` in that case. We can return the genesis information from the pallet for the client to use. Will you be able to pick this?

I'm not sure how to do this: 
1. it seems that modifying `GenesisData` implies modifying `genesis.json` (because this file is deserialized into a GenesisData object),
2. `GenesisData` itself is only used to get a `GenesisConfig` object, so that would imply modifying the definition of `GenesisConfig`,
3. `GenesisConfig` is used only through the `build()` method, which writes stuff to the pallet storage. So I assume there should be another `pallet::storage` for predeployed accounts in `pallets/starknet/src/lib.rs` ? Otherwise, maybe we can use a `pallet::constant` in the definition of `Config`, but idk how one should write constants to the config when building GenesisConfig.

tdelabro replied:
I'm fine with it. The genesis.json fine is copied under the chain data directory, so it should stay there an not be touched or removed during the whole time the chain run

Juul-Mc-Goa commented:
@apoorvsadana  I changed the rpc method to make it read the `genesis.json` file.

I assumed there should be as little hardcoded constants as possible, so now:
1. the `PredeployedAccount` struct only stores the address, contract class, and balance (so no `private_key` attribute),
2. the `private_key` attribute is now part of the struct `PredeployedAccountsInfo` as I found no easy way to find wether an account is NoValidate or Argent (that is, without hardcoding the class hashes of the relevant contracts),
3. there is now only 2 constants: 
    - `GENESIS_FILE`: the path to `genesis.json`,
    - `ARGENT_PK`: the private key, which (I believe) appears nowhere in the `genesis.json` file.

I feel it would be more natural if we could change the genesis file however, along with the `GenesisData/GenesisConfig` structs and `Config` trait in the pallet. This way we could at least store the private key in the genesis file instead of hardcoding it. And we could also easily store which account is NoValidate. We could even store the predeployed accounts and the pk as part of the pallet's `Config` trait. I'm too much of a beginner to say if that's a good idea though.

tdelabro commented on crates/client/rpc-core/src/lib.rs:
```diff
@@ -27,9 +27,26 @@ use starknet_core::types::{
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct PredeployedAccountsInfo {
+    pub accounts: Vec<PredeployedAccount>,
+    pub private_key: FieldElement,
+}
+
```
Why is there a single private_key when multiple accounts can be deployed?
I would expect the possibility to fill in one for each account

Juul-Mc-Goa replied:
I can change that np

tdelabro commented on crates/client/rpc-core/src/lib.rs:
```diff
@@ -27,9 +27,26 @@ use starknet_core::types::{
+#[derive(Serialize, Deserialize)]
+pub struct PredeployedAccountsInfo {
+    pub accounts: Vec<PredeployedAccount>,
+    pub private_key: FieldElement,
```
private_key does not have to be a FieldElement, it can be anything. So I would have it be a `Vec<u8>`

Juul-Mc-Goa replied:
alright

tdelabro commented on crates/client/rpc/src/constants.rs:
```diff
@@ -2,3 +2,9 @@
+pub const GENESIS_FILE: &str = "configs/genesis-assets/genesis.json";
+
+/// Pk for the relevant predeployed accounts defined in the genesis state.
+pub const ARGENT_PK: &str = "0x00c1cf1490de1352865301bb8705143f3ef938f97fdf892f1090dcb5ac7bcd1d";
```
ARGENT_PK should not be hardcoded.
We discussed it here: https://github.com/keep-starknet-strange/madara/pull/1265#discussion_r1402066121. You have to make it configurable in genesis.json by adding an extra field to the struct  with the following tuple (address, name, pk)

Juul-Mc-Goa replied:
ah ok so I do have to modify genesis.json, sry I had misunderstood.

tdelabro commented on crates/client/rpc/src/constants.rs:
```diff
@@ -2,3 +2,9 @@
 pub const MAX_EVENTS_CHUNK_SIZE: usize = 1000;
+
+/// Path to the `genesis.json` file
+pub const GENESIS_FILE: &str = "configs/genesis-assets/genesis.json";
```
This should not be hardcoded. And it should not be treated from this path.
Use `load_genesis` fn defined in ` crates/node/src/chain_spec.rs`

Juul-Mc-Goa replied:
`crates/node/src/chain_spec.rs` cannot be used in `crates/client/...` because it would result in cyclic dependencies. That's why I basically copy/pasted it. Maybe it can be moved to another file however, and `use`d in both places.

apoorvsadana replied:
Ya I agree with moving it to a utils crate - https://github.com/keep-starknet-strange/madara/pull/1265#discussion_r1404457326

apoorvsadana commented on crates/client/rpc/src/lib.rs:
```diff
@@ -98,6 +101,19 @@ where
         Ok(UniqueSaturatedInto::<u64>::unique_saturated_into(self.client.info().best_number))
     }
+
+    fn load_genesis() -> GenesisData {
+        let genesis_path: PathBuf = String::from(GENESIS_FILE).into();
+        let genesis_file_content = std::fs::read_to_string(genesis_path.clone()).unwrap_or_else(|_| {
+            panic!(
+                "Failed to read genesis file at {}. Please run `madara setup` before opening an issue.",
+                genesis_path.canonicalize().unwrap().display()
+            )
+        });
+        let genesis_data: GenesisData = serde_json::from_str(&genesis_file_content).expect("Failed loading genesis");
+
+        genesis_data
+    }
```
I guess this is the same function used inside the node crate. As @tdelabro mentioned, let's reuse that.

tdelabro replied:
And it use a dynamic path rather than the one you hardcoded

apoorvsadana replied:
Ya we should do that imo. I think we can have a utils crate at the same level as `node`, `client` etc. for functions that are used across crates. I am not sure if this is the idiomatic approach, @tdelabro?

apoorvsadana replied:
Awesome, @Juul-Mc-Goa can you make this change?

Juul-Mc-Goa replied:
yep working on it

apoorvsadana commented on crates/client/rpc-core/src/lib.rs:
```diff
@@ -27,9 +27,26 @@ use starknet_core::types::{
+#[derive(Serialize, Deserialize)]
+pub struct PredeployedAccount {
+    pub contract_address: FieldElement,
+    pub contract_class: ContractClass,
```
Is there any specific reason to return the `contract_class`? I am not sure how it will be used by the client and it make the response really big.

tdelabro replied:
You can return the ContractClassHash. It's informative and light

Juul-Mc-Goa replied:
No real reason, I wasn't aware the response would be that big until I tested it. I can definitely replace this attribute with a class hash.

Juul-Mc-Goa commented:
@tdelabro @apoorvsadana  OK so I changed the genesis assets file, changed `PredeployedAccount` and `GenesisData` accordingly.

HOWEVER: I cannot `use` the `node` crate inside the `client` crate due to cyclic dependencies, so there is still a `load_genesis` function in `crates/client/rpc/src/lib.rs`, and a [hardcoded constant to find the file](https://github.com/keep-starknet-strange/madara/pull/1265#discussion_r1404217462).

tdelabro commented on crates/client/rpc/src/lib.rs:
```diff
@@ -98,6 +102,23 @@ where
     }
+
+    fn load_genesis() -> GenesisData {
+        let base_path: PathBuf = BasePath::from_project("", "", "madara").config_dir("dev");
```
A more safe way would be to create a `GenesisProvider` trait, and to make all those structs generic over a type that implemnt this trait.

This way the RPC just have to call the `load_genesis_data` method defined by this trait, and don't have to worry about it being store on disk or being hardcoded or anything.
It just call the method and get a `Result<GenesisData, LoadGenesisDataError>`.

On the other hand. You impl this trait in a `struct OnDiskGenesisConfig<'a>(&'a Path)`, and you pass this struct as argument to all the struct we talked about previously.

Juul-Mc-Goa commented on crates/client/rpc/src/constants.rs:
```diff
@@ -2,3 +2,6 @@
 pub const MAX_EVENTS_CHUNK_SIZE: usize = 1000;
+
+/// Path to the genesis assets file
+pub const GENESIS_ASSETS_PATH: &[&str] = &["genesis-assets", "genesis.json"];
```
yeah ok I thought it'd be better to use `PathBuf::join()` rather than using the unix "/" delimiter

tdelabro replied:
yes you are right
